test/unit/uri.c

```diff
@@ -590,16 +597,106 @@ test_invalid_string_uri_set(void)
 	footer();
 }
 
+#define RFC3986_unreserved "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._~"
+
+const struct str_escape escape_testcase[] = {
+	[0] = {
+		.raw = "-._~",
+		.escaped = "-._~",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[1] = {
+		.raw = "0123456789",
+		.escaped = "0123456789",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[2] = {
+		.raw = "abcdefghijklm",
+		.escaped = "abcdefghijklm",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[3] = {
+		.raw = "nopqrstuvwxyz",
+		.escaped = "nopqrstuvwxyz",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[4] = {
+		.raw = "ABCDEFGHIJKLM",
+		.escaped = "ABCDEFGHIJKLM",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[5] = {
+		.raw = "NOPQRSTUVWXYZ",
+		.escaped = "NOPQRSTUVWXYZ",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[6] = {
+		.raw = "!$&'()*+,;=",
+		.escaped = "%21%24%26%27%28%29%2A%2B%2C%3B%3D",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+};
+
+static void
+test_escape(void)
+{
+	header();
+	plan(lengthof(escape_testcase) * 2);
+	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
+		const char *raw = escape_testcase[i].raw;
+		const char *escaped = escape_testcase[i].escaped;
+		const char *unreserved = escape_testcase[i].unreserved;
+		bool plus = escape_testcase[i].plus;
+		char *enc = uri_escape(raw, strlen(raw), unreserved, plus);
+		isnt(strlen(enc), 0, "escaped string ('%s') length != 0",
+		     enc);
+		is(strcmp(escaped, enc), 0, "escape: '%s' == '%s'",
+		   escaped, enc);
+		free(enc);
+	}
+	check_plan();
+	footer();
+}
+
+static void
+test_unescape(void)
+{
+	header();
+	plan(lengthof(escape_testcase) * 2);
+	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
+		const char *raw = escape_testcase[i].raw;
+		const char *escaped = escape_testcase[i].escaped;
+		bool decode_plus = escape_testcase[i].plus;
+		char *dec = uri_unescape((char *)escaped, decode_plus);
```

Comment:
It casts a value from `const char *` to `char *` to pass it as `const char *` argument to a function. The cast is unnecessary.